### ISSUE TITLE
"fuse" options to consolidate heterogeneous type definitions

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#2852]
Repro is with Zed commit 2b0b0c9.

A community user spawned a discussion regarding what happens when `fuse` is applied on inputs that would normally create multiple type definitions for values stored in fields of the same name.

Consider the following NDJSON sample data:

```
$ cat example.ndjson 
{"tags": []}
{"tags": ["something"]}
{"num": 1}
{"num": 2.1}

$ zq -version
Version: v0.29.0-411-g2b0b0c97

$ zq -Z 'put typedef := typeof(this)' example.ndjson 
{
    tags: [] (0=([null])),
    typedef: ({tags:[null]})
}
{
    tags: [
        "something"
    ],
    typedef: ({tags:[string]})
}
{
    num: 1,
    typedef: ({num:int64})
}
{
    num: 2.1,
    typedef: ({num:float64})
}
```

Maybe the user's intent was that `tags` would _always_ be treated as an array of strings and `num` would _always_ be treated as floating point. But since no shaper was applied, the behavior shown all makes sense.

When `fuse` is applied, the heterogeneous type definitions have the effect of automatically creating separate named fields for each variation by appending `_2` to the original field names.

```
$ zq -Z 'fuse' example.ndjson 
{
    tags: [] (0=([null])),
    tags_2: null (1=([string])),
    num: null (int64),
    num_2: null (float64)
}
{
    tags: null (0),
    tags_2: [
        "something"
    ],
    num: null (int64),
    num_2: null (float64)
}
{
    tags: null (0),
    tags_2: null (1),
    num: 1,
    num_2: null (float64)
}
{
    tags: null (0),
    tags_2: null (1),
    num: null (int64),
    num_2: 2.1
}
```

This may be fine for the needs of some users, but it threw this particular community user for a loop, so it's worth considering if we might add one or more options to vary the behavior. Two thoughts came out of the discussion:

1.  (My idea :smile: ) As `fuse` is completing its first pass through the data, perhaps it could track if all the observed type definitions for a named field would fit under just one of them. For instance, the empty array for `tags` could be stored using the type definition for the array of strings, and likewise the numeric value `1` could be stored as a float.
2.  (The user's idea :smile: ) Have `fuse` create a `union` type that covers all the observed values. So for example, `tags` could have been an array of union (string, int) in the fused type.

FWIW, after seeing them laid out side-by-side, the community user offered the following take:

> I think I prefer \[option 1\] when it’s possible to fit all the data under the same type (e.g., \[\] and \[“something”\]), and then \[option 2\] when that isn’t possible (e.g., \[1.2\] and \[“something”\]).

**Comment by philrz** [comment]
While going back over old PRs to compile some release notes, I recognized that the `shape` operator added via #2327 seems somewhat related to this topic.

**Comment by nwt** [comment]
Here's the output for #2885.
```
$ go run ./cmd/zq -Z fuse example.ndjson
{
    tags: [] (0=([null])) (1=((0,2=([string])))),
    num: null (3=((int64,float64)))
} (=4)
{
    tags: [
        "something"
    ],
    num: null
} (4)
{
    tags: null,
    num: 1
} (4)
{
    tags: null,
    num: 2.1
} (4)
```

The `tags` field is now a union, `(([null],[string]))`. We can simplify that to a non-union, `[string]`, as suggested above, but then we need to decide whether to keep or drop `[null]` in a union like `(([null],[string],[int64]))`.

**Comment by mccanne** [comment]
The idea to name fields with _k suffixes seems grungy as a feature of fuse.  Noah and I agree we should have another operator that unflattens union values that you could apply to the output of fuse if you'd like.  As Phil suggested, when there are two union type and one of them is null, we should just use the non-null type without a union.

**Comment by mccanne** [comment]
This is partially fixed and now generates:
```
tags:[]([null])(([null],[string])),num:null((int64,float64))}
{tags:["something"](([null],[string])),num:null((int64,float64))}
{tags:null(([null],[string])),num:1((int64,float64))}
{tags:null(([null],[string])),num:2.1((int64,float64))}
```
When there is a union of two types and one is a null, it should be just the one non-null type though this case is complicated by the fact that the null is [null]...

**Comment by philrz** [comment]
I revisited this issue while doing some scrubbing of shaping-related issues, and based on a binary search it looks like this was fixed thanks to #3717. As suggested in @mccanne's comment above, now the `[null]` is no longer present. Here's how the output looks as of commit 12ab2c9 that's associated with #3717, and it's been the same since.

```
$ zq -version
Version: v0.33.0-282-g12ab2c91

$ zq -z 'fuse' example.ndjson
{tags:[]([string]),num:null((int64,float64))}
{tags:["something"],num:null((int64,float64))}
{tags:null([string]),num:1((int64,float64))}
{tags:null([string]),num:2.1((int64,float64))}
```

I'll leave it to the Dev team to do the final review & close here.
